File: tests/mandelbrot.rb
#!/usr/bin/env ruby

def mandelbrot(h, w, max_iter)
  # Define the region of the complex plane to plot
  y_min, y_max = -1.2, 1.2
  x_min, x_max = -2.0, 1.0
  
  # Initialize the output array
  result = Array.new(h) { Array.new(w, 0) }
  
  # For each pixel in the output image
  (0...h).each do |i|
    (0...w).each do |j|
      # Convert pixel coordinates to complex plane coordinates
      x = x_min + j * (x_max - x_min) / (w - 1)
      y = y_min + i * (y_max - y_min) / (h - 1)
      
      # The complex number to test
      c_real, c_imag = x, y
      z_real, z_imag = 0.0, 0.0
      
      # Iterate to see if the point escapes
      max_iter.times do |k|
        # Calculate z = z^2 + c manually without Complex class
        # (a+bi)^2 = a^2 - b^2 + 2abi
        z_real_temp = z_real*z_real - z_imag*z_imag + c_real
        z_imag = 2*z_real*z_imag + c_imag
        z_real = z_real_temp
        
        # If the point escapes (|z| > 2)
        if z_real*z_real + z_imag*z_imag > 4
          # Color is based on how quickly the point escapes
          result[i][j] = k + 1
          break
        end
      end
      
      # If the point doesn't escape, it's in the Mandelbrot set
      # We leave it as 0 in the result
    end
  end
  
  return result
end

def print_mandelbrot(mandelbrot_set)
  h = mandelbrot_set.length
  w = mandelbrot_set[0].length
  
  # Define a set of characters with increasing "density"
  chars = ' .:-=+*#%@'
  
  # Find the maximum value in the set for normalization
  max_value = mandelbrot_set.flatten.max
  
  (0...h).each do |i|
    (0...w).each do |j|
      if mandelbrot_set[i][j] == 0
        # Points in the set
        print '@'
      else
        # Points outside the set, with a character representing how quickly they escape
        index = (chars.length - 1) * mandelbrot_set[i][j] / max_value
        print chars[index.to_i]
      end
    end
    puts  # New line after each row
  end
end

# Set parameters
height = 120
width = 100
max_iterations = 1000

# Generate and print the Mandelbrot set
puts "Generating Mandelbrot set..."
mandelbrot_set = mandelbrot(height, width, max_iterations)
puts "Printing Mandelbrot set:"
print_mandelbrot(mandelbrot_set)

# For a better view, let's also save the output to a file
puts "\nSaving output to 'mandelbrot.txt'..."
File.open('mandelbrot.txt', 'w') do |f|
  chars = ' .:-=+*#%@'
  max_value = mandelbrot_set.flatten.max
  
  (0...height).each do |i|
    (0...width).each do |j|
      if mandelbrot_set[i][j] == 0
        f.write('@')
      else
        index = (chars.length - 1) * mandelbrot_set[i][j] / max_value
        f.write(chars[index.to_i])
      end
    end
    f.write("\n")
  end
end

puts "Done!"

File: tests/mandelbrot.py
def mandelbrot(h, w, max_iter):
    """
    Returns a list of lists representing the Mandelbrot set.
    h: height of the output image
    w: width of the output image
    max_iter: maximum number of iterations to check for convergence
    """
    # Define the region of the complex plane to plot
    y_min, y_max = -1.2, 1.2
    x_min, x_max = -2.0, 1.0
    
    # Initialize the output array
    result = [[0 for _ in range(w)] for _ in range(h)]
    
    # For each pixel in the output image
    for i in range(h):
        for j in range(w):
            # Convert pixel coordinates to complex plane coordinates
            x = x_min + j * (x_max - x_min) / (w - 1)
            y = y_min + i * (y_max - y_min) / (h - 1)
            
            # The complex number to test
            c = complex(x, y)
            z = complex(0, 0)
            
            # Iterate to see if the point escapes
            for k in range(max_iter):
                z = z * z + c
                
                # If the point escapes
                if abs(z) > 2:
                    # Color is based on how quickly the point escapes
                    result[i][j] = k
                    break
            
            # If the point doesn't escape, it's in the Mandelbrot set
            # We leave it as 0 in the result
    
    return result

def print_mandelbrot(mandelbrot_set):
    """
    Print a simple ASCII representation of the Mandelbrot set.
    """
    h = len(mandelbrot_set)
    w = len(mandelbrot_set[0])
    
    # Define a set of characters with increasing "density"
    chars = ' .:-=+*#%@'
    max_value = max(max(row) for row in mandelbrot_set)
    
    for i in range(h):
        for j in range(w):
            # Scale the value to the range of characters
            if mandelbrot_set[i][j] == 0:
                # Points in the set
                print('@', end='')
            else:
                # Points outside the set, with a character representing how quickly they escape
                index = (len(chars) - 1) * mandelbrot_set[i][j] // max_value
                print(chars[index], end='')
        print()  # New line after each row

# Set parameters
height = 100
width = 120
max_iterations = 1000

# Generate and print the Mandelbrot set
mandelbrot_set = mandelbrot(height, width, max_iterations)
print_mandelbrot(mandelbrot_set)

# For a better view, let's also save the output to a file
with open('mandelbrot.txt', 'w') as f:
    chars = ' .:-=+*#%@'
    max_value = max(max(row) for row in mandelbrot_set)
    
    for i in range(height):
        for j in range(width):
            if mandelbrot_set[i][j] == 0:
                f.write('@')
            else:
                index = (len(chars) - 1) * mandelbrot_set[i][j] // max_value
                f.write(chars[index])
        f.write('\n')

print("\nOutput also saved to 'mandelbrot.txt'")
